File: frontend/src/spread/utm.ts
/**
 * UTM ↔ WGS84, hand-rolled (no proj4). The inverse is the standard Snyder
 * series on the WGS84 ellipsoid (sub-meter inside a zone — far below the
 * 30 m pixel size of the archive grids), used to corner-pin decoded rasters
 * onto MapLibre canvas sources. The forward direction (lonLatToUtm) is the
 * Krüger n-series, mm-level, mirroring worker/responder_worker/utm.py so a
 * Walk pin lands in the same routing-grid cell the worker computed.
 */

const A = 6378137; // WGS84 semi-major axis
const F = 1 / 298.257223563;
const K0 = 0.9996;
const E0 = 500000; // false easting
const N0_SOUTH = 10000000; // false northing, southern hemisphere

const E2 = F * (2 - F); // first eccentricity squared
const EP2 = E2 / (1 - E2); // second eccentricity squared
const E1 = (1 - Math.sqrt(1 - E2)) / (1 + Math.sqrt(1 - E2));

/** EPSG 326xx/327xx → zone + hemisphere; null for anything else. */
export function epsgToUtm(epsg: number): { zone: number; northern: boolean } | null {
  const base = Math.floor(epsg / 100);
  const zone = epsg % 100;
  if (zone < 1 || zone > 60) return null;
  if (base === 326) return { zone, northern: true };
  if (base === 327) return { zone, northern: false };
  return null;
}

/** Central meridian of a UTM zone, degrees. */
export function zoneCentralMeridian(zone: number): number {
  return zone * 6 - 183;
}

/** UTM easting/northing (meters) → [lon, lat] degrees (WGS84). */
export function utmToLonLat(
  easting: number,
  northing: number,
  zone: number,
  northern = true,
): [number, number] {
  const x = easting - E0;
  const y = northern ? northing : northing - N0_SOUTH;

  // Footpoint latitude from the meridian arc.
  const m = y / K0;
  const mu = m / (A * (1 - E2 / 4 - (3 * E2 * E2) / 64 - (5 * E2 * E2 * E2) / 256));
  const phi1 =
    mu +
    ((3 * E1) / 2 - (27 * E1 ** 3) / 32) * Math.sin(2 * mu) +
    ((21 * E1 ** 2) / 16 - (55 * E1 ** 4) / 32) * Math.sin(4 * mu) +
    ((151 * E1 ** 3) / 96) * Math.sin(6 * mu) +
    ((1097 * E1 ** 4) / 512) * Math.sin(8 * mu);

  const sinPhi1 = Math.sin(phi1);
  const cosPhi1 = Math.cos(phi1);
  const tanPhi1 = Math.tan(phi1);

  const c1 = EP2 * cosPhi1 * cosPhi1;
  const t1 = tanPhi1 * tanPhi1;
  const oneMinusE2Sin2 = 1 - E2 * sinPhi1 * sinPhi1;
  const n1 = A / Math.sqrt(oneMinusE2Sin2);
  const r1 = (A * (1 - E2)) / Math.pow(oneMinusE2Sin2, 1.5);
  const d = x / (n1 * K0);

  const lat =
    phi1 -
    ((n1 * tanPhi1) / r1) *
      ((d * d) / 2 -
        ((5 + 3 * t1 + 10 * c1 - 4 * c1 * c1 - 9 * EP2) * d ** 4) / 24 +
        ((61 + 90 * t1 + 298 * c1 + 45 * t1 * t1 - 252 * EP2 - 3 * c1 * c1) * d ** 6) / 720);
  const lon =
    (d -
      ((1 + 2 * t1 + c1) * d ** 3) / 6 +
      ((5 - 2 * c1 + 28 * t1 - 3 * c1 * c1 + 8 * EP2 + 24 * t1 * t1) * d ** 5) / 120) /
    cosPhi1;

  const deg = 180 / Math.PI;
  return [zoneCentralMeridian(zone) + lon * deg, lat * deg];
}

/** MapLibre image/canvas source corner order: TL, TR, BR, BL ([lon, lat]). */
export type Corners = [
  [number, number],
  [number, number],
  [number, number],
  [number, number],
];

/**
 * Convert a UTM bbox ([minX, minY, maxX, maxY], geotiff getBoundingBox order)
 * to WGS84: all 4 corners inverted independently (the projected rectangle is
 * NOT axis-aligned in lon/lat) → MapLibre corner-pin coordinates + the
 * enclosing [w, s, e, n] bounds.
 */
export function utmBoundsTo4326(
  bbox: [number, number, number, number],
  zone: number,
  northern = true,
): { corners: Corners; bounds: [number, number, number, number] } {
  const [minX, minY, maxX, maxY] = bbox;
  const tl = utmToLonLat(minX, maxY, zone, northern);
  const tr = utmToLonLat(maxX, maxY, zone, northern);
  const br = utmToLonLat(maxX, minY, zone, northern);
  const bl = utmToLonLat(minX, minY, zone, northern);
  const corners: Corners = [tl, tr, br, bl];
  const lons = corners.map((c) => c[0]);
  const lats = corners.map((c) => c[1]);
  return {
    corners,
    bounds: [Math.min(...lons), Math.min(...lats), Math.max(...lons), Math.max(...lats)],
  };
}

// Krüger n-series (4th order) constants for the forward projection.
const KN = F / (2 - F);
const KA = (A / (1 + KN)) * (1 + KN ** 2 / 4 + KN ** 4 / 64);
const ALPHA = [
  KN / 2 - (2 * KN ** 2) / 3 + (5 * KN ** 3) / 16 + (41 * KN ** 4) / 180,
  (13 * KN ** 2) / 48 - (3 * KN ** 3) / 5 + (557 * KN ** 4) / 1440,
  (61 * KN ** 3) / 240 - (103 * KN ** 4) / 140,
  (49561 * KN ** 4) / 161280,
];
const KC = (2 * Math.sqrt(KN)) / (1 + KN);

/** WGS84 UTM zone for a longitude (no Norway/Svalbard exceptions; CONUS). */
export function utmZoneFor(lon: number): number {
  return Math.min(60, Math.max(1, Math.floor((lon + 180) / 6) + 1));
}

/** [lon, lat] degrees (WGS84) → UTM [easting, northing] metres. */
export function lonLatToUtm(
  lon: number,
  lat: number,
  zone: number,
  northern = true,
): [number, number] {
  const phi = (lat * Math.PI) / 180;
  const lam = ((lon - zoneCentralMeridian(zone)) * Math.PI) / 180;
  const sphi = Math.sin(phi);
  const t = Math.sinh(Math.atanh(sphi) - KC * Math.atanh(KC * sphi));
  const xiP = Math.atan2(t, Math.cos(lam));
  const etaP = Math.atanh(Math.sin(lam) / Math.sqrt(1 + t * t));
  let xi = xiP;
  let eta = etaP;
  for (let j = 1; j <= 4; j++) {
    const a = ALPHA[j - 1];
    xi += a * Math.sin(2 * j * xiP) * Math.cosh(2 * j * etaP);
    eta += a * Math.cos(2 * j * xiP) * Math.sinh(2 * j * etaP);
  }
  return [E0 + K0 * KA * eta, K0 * KA * xi + (northern ? 0 : N0_SOUTH)];
}

File: worker/responder_worker/utm.py
"""WGS84 <-> UTM, Krüger n-series to 4th order (mm-level inside a zone).

Pure numpy, vectorized: every function takes scalars or arrays. The routing
bundle builder uses it for AOI boxes, OSM node projection and the graph
coordinates; the browser mirrors the forward direction in
frontend/src/spread/utm.ts so A/B pins land in the same grid cells.
"""

from __future__ import annotations

import math

import numpy as np

A_AXIS = 6378137.0
F = 1 / 298.257223563
K0 = 0.9996
E0 = 500000.0
N0_SOUTH = 10000000.0

_n = F / (2 - F)
_A = A_AXIS / (1 + _n) * (1 + _n**2 / 4 + _n**4 / 64)
_ALPHA = (
    _n / 2 - 2 * _n**2 / 3 + 5 * _n**3 / 16 + 41 * _n**4 / 180,
    13 * _n**2 / 48 - 3 * _n**3 / 5 + 557 * _n**4 / 1440,
    61 * _n**3 / 240 - 103 * _n**4 / 140,
    49561 * _n**4 / 161280,
)
_BETA = (
    _n / 2 - 2 * _n**2 / 3 + 37 * _n**3 / 96 - _n**4 / 360,
    _n**2 / 48 + _n**3 / 15 - 437 * _n**4 / 1440,
    17 * _n**3 / 480 - 37 * _n**4 / 840,
    4397 * _n**4 / 161280,
)
_DELTA = (
    2 * _n - 2 * _n**2 / 3 - 2 * _n**3 + 116 * _n**4 / 45,
    7 * _n**2 / 3 - 8 * _n**3 / 5 - 227 * _n**4 / 45,
    56 * _n**3 / 15 - 136 * _n**4 / 35,
    4279 * _n**4 / 630,
)
_C = 2 * math.sqrt(_n) / (1 + _n)


def zone_for(lon: float) -> int:
    return int(min(60, max(1, math.floor((lon + 180) / 6) + 1)))


def epsg_for(lon: float, lat: float) -> int:
    return (32600 if lat >= 0 else 32700) + zone_for(lon)


def zone_of_epsg(epsg: int) -> tuple[int, bool]:
    base, zone = divmod(int(epsg), 100)
    if base not in (326, 327) or not 1 <= zone <= 60:
        raise ValueError(f"not a WGS84 UTM epsg: {epsg}")
    return zone, base == 326


def central_meridian(zone: int) -> float:
    return zone * 6 - 183.0


def fwd(lon, lat, zone: int, northern: bool = True):
    """lon/lat degrees -> (easting, northing) metres."""
    phi = np.radians(np.asarray(lat, dtype=np.float64))
    lam = np.radians(np.asarray(lon, dtype=np.float64) - central_meridian(zone))
    sphi = np.sin(phi)
    t = np.sinh(np.arctanh(sphi) - _C * np.arctanh(_C * sphi))
    xi_p = np.arctan2(t, np.cos(lam))
    eta_p = np.arctanh(np.sin(lam) / np.sqrt(1 + t * t))
    xi, eta = xi_p.copy(), eta_p.copy()
    for j, a in enumerate(_ALPHA, start=1):
        xi = xi + a * np.sin(2 * j * xi_p) * np.cosh(2 * j * eta_p)
        eta = eta + a * np.cos(2 * j * xi_p) * np.sinh(2 * j * eta_p)
    e = E0 + K0 * _A * eta
    n = K0 * _A * xi + (0.0 if northern else N0_SOUTH)
    if np.ndim(e) == 0:
        return float(e), float(n)
    return e, n


def inv(easting, northing, zone: int, northern: bool = True):
    """(easting, northing) metres -> (lon, lat) degrees."""
    e = np.asarray(easting, dtype=np.float64)
    n = np.asarray(northing, dtype=np.float64) - (0.0 if northern else N0_SOUTH)
    xi = n / (K0 * _A)
    eta = (e - E0) / (K0 * _A)
    xi_p, eta_p = xi.copy(), eta.copy()
    for j, b in enumerate(_BETA, start=1):
        xi_p = xi_p - b * np.sin(2 * j * xi) * np.cosh(2 * j * eta)
        eta_p = eta_p - b * np.cos(2 * j * xi) * np.sinh(2 * j * eta)
    chi = np.arcsin(np.sin(xi_p) / np.cosh(eta_p))
    phi = chi.copy()
    for j, d in enumerate(_DELTA, start=1):
        phi = phi + d * np.sin(2 * j * chi)
    lam = np.arctan2(np.sinh(eta_p), np.cos(xi_p))
    lon = central_meridian(zone) + np.degrees(lam)
    lat = np.degrees(phi)
    if np.ndim(lon) == 0:
        return float(lon), float(lat)
    return lon, lat


def bbox_lonlat_to_utm(bbox4326, zone: int, northern: bool = True, samples: int = 8):
    """Enclosing UTM box of a lon/lat box (edges densified: the projected
    rectangle is curved, so corners alone under-cover it)."""
    w, s, e, nn = bbox4326
    ts = np.linspace(0.0, 1.0, samples + 1)
    lons = np.concatenate([w + (e - w) * ts, np.full_like(ts, e), w + (e - w) * ts, np.full_like(ts, w)])
    lats = np.concatenate([np.full_like(ts, s), s + (nn - s) * ts, np.full_like(ts, nn), s + (nn - s) * ts])
    x, y = fwd(lons, lats, zone, northern)
    return (float(x.min()), float(y.min()), float(x.max()), float(y.max()))


def bbox_utm_to_lonlat(bbox_utm, zone: int, northern: bool = True, samples: int = 8):
    """Enclosing lon/lat box of a UTM rectangle (edges densified)."""
    x0, y0, x1, y1 = bbox_utm
    ts = np.linspace(0.0, 1.0, samples + 1)
    xs = np.concatenate([x0 + (x1 - x0) * ts, np.full_like(ts, x1), x0 + (x1 - x0) * ts, np.full_like(ts, x0)])
    ys = np.concatenate([np.full_like(ts, y0), y0 + (y1 - y0) * ts, np.full_like(ts, y1), y0 + (y1 - y0) * ts])
    lon, lat = inv(xs, ys, zone, northern)
    return (float(lon.min()), float(lat.min()), float(lon.max()), float(lat.max()))
